src/ssh/cockpitsshrelay.c

```diff
@@ -896,96 +896,175 @@ do_password_auth (CockpitSshData *data)
   return rc;
 }
 
+struct CockpitSshPromptData {
+  CockpitSshData *data;
+  const gchar *identity;
+  gboolean did_prompt;
+};
+
 /* We don't support unlocking identities within cockpit-ssh so fail here */
 static int
 prompt_for_identity_password (const char *prompt, char *buf, size_t len,
                               int echo, int verify, void *userdata)
 {
-  CockpitSshData *data = (CockpitSshData *) userdata;
-  gchar *identity = NULL;
-  if (ssh_options_get (data->session, SSH_OPTIONS_IDENTITY, &identity) == SSH_OK)
-    {
-      data->problem_error = g_strdup_printf ("locked identity: %s", identity);
-    }
-  if (identity)
-    ssh_string_free_char (identity);
+  struct CockpitSshPromptData *prompt_data = userdata;
+  prompt_data->data->problem_error = g_strdup_printf ("locked identity: %s", prompt_data->identity);
+  prompt_data->did_prompt = TRUE;
   return -1;
 }
 
 static int
-do_key_identity_auth (CockpitSshData *data)
+do_auto_auth (CockpitSshData *data)
 {
+  /* HACK: When prompting for a key passphrase, libssh doesn't provide
+     enough information to say which key it is for, but we need that
+     information since Cockpit will offer to load the key into the
+     agent in order to log in.
+
+     To get this information, we reimplement
+     ssh_userauth_publickey_auto here in a way that passes that
+     information to the callback.
+
+     We would like to iterate over all configured identities, the same
+     way that the real ssh_userauth_publickey does, but there is no
+     API to do that either.  So we hard code all the names, based on
+     what ssh-add would add to the agent.
+
+     This whole thing is meant to be replaced with something like this
+     eventually:
+
+     static int
+     prompt_for_identity_password (const char *prompt, char *buf, size_t len,
+                                   int echo, int verify, void *userdata)
+     {
+       struct CockpitSshData *data = userdata;
+       const gchar *identity = ssh_userauth_publickey_auto_get_current_identity (data->session);
```

Comment:
Oh this would be neat :)

Reply:
https://gitlab.com/libssh/libssh-mirror/-/merge_requests/134 :-)